runtime/compiler/trj9/runtime/ClassUnloadAssumption.cpp

```diff
@@ -351,6 +354,149 @@ TR_RuntimeAssumptionTable::detachFromRAT(OMR::RuntimeAssumption *assumption)
+                     {
+                     TR_ASSERT( hashTable->_htSpineArray[i] == cursor, "RAT spine head is not cursor!" );
+                     hashTable->_htSpineArray[i] = next;
+                     }
+                  hashTable->_markedforDetachCount[i]--;
+                  // Now release the assumption
+                  incReclaimedAssumptionCount(kind);
+                  cursor->reclaim();
+                  cursor->paint(); // RAS
+                  TR_PersistentMemory::jitPersistentFree(cursor);
+                  reclaimed++; // RAS
+                  }
+               else
+                  {
+                  prev = cursor;
+                  }
+               }
+            TR_ASSERT( hashTable->_markedforDetachCount[i]==0, "RAT detach count should be 0!" );
+            }
+         _detachPending[kind] = false;
+         }
+      }
+   assumptionTableMutex->exit();
+   TR_ASSERT( marked == reclaimed, "Did not reclaime all the marked assumptions!" );
+   marked=0;
+   }
+
+/**
+ * Mark and detach all of the assumptions in the metadata's circular linked list
+ * @param md The metadata for which to mark & detach assumptions from.
+ * @param reclaimPrePrologueAssumptions Are all assumptions free to be removed (ClassUnloading vs. recompiled)
+ * This only detaches from the metadata's linked list, not a detech from the RAT,
+ * to remove from the RAT you need to call reclaimMarkedAssumptionsFromRAT()
+ */
+void TR_RuntimeAssumptionTable::markAssumptionsAndDetach(void * md, bool reclaimPrePrologueAssumptions)
+   {
+   J9JITExceptionTable *metaData = (J9JITExceptionTable*) md;
+   OMR::RuntimeAssumption *sentry = (OMR::RuntimeAssumption*)(metaData->runtimeAssumptionList);
+   OMR::RuntimeAssumption *cursor, *next;
+
+   assumptionTableMutex->enter();
+   if (sentry)
+      {
+      int32_t numAssumptionsNotReclaimed = 0; // RAS
+      TR_ASSERT(sentry->getAssumptionKind() == RuntimeAssumptionSentinel, "First assumption must be the sentinel\n");
+      OMR::RuntimeAssumption *notReclaimedList = sentry;
+      for ( cursor=sentry->getNextAssumptionForSameJittedBody(); cursor != sentry; cursor=next)
+         {
+         next = cursor->getNextAssumptionForSameJittedBody();
+         if (!metaData || cursor->isAssumingMethod(metaData, reclaimPrePrologueAssumptions))
+            {
+            markForDetachFromRAT(cursor); // Mark for deletion
+            marked++;
+            }
+         else // This could be an assumption to the persistentBodyInfo which is not reclaimed
+            {
+            cursor->setNextAssumptionForSameJittedBody(notReclaimedList);
+            notReclaimedList = cursor;
+            numAssumptionsNotReclaimed++; // RAS
+            }
+         }
+
+      if (notReclaimedList != sentry) // some entries were not reclaimed
+         {
+         // Must attach the notReclaimedList to the sentinel
+         sentry->setNextAssumptionForSameJittedBody(notReclaimedList);
+
+#ifdef PROD_WITH_ASSUMES
```

Comment:
I think this may want to be #if defined(PROD_WITH_ASSUMES) || defined(DEBUG) since we want this code in either case. Perhaps this should also be available under an option or env var for prod builds to help with bug finding later?